### ISSUE TITLE
keras metrics discards imaginary part of complex input and casts to float by taking the real part when passing complex numbers to metric function

### ISSUE THREAD

**Issue by isaacgerg** [keras-team/keras#15737]
TF 2.4.1 - 2.7

When using a metric that accepts complex numbers, tf keras fit ends up discarding the imaginary part of the complex number and then casts the array to real when inputting y_pred and y_true to a metric function.

It works correctly when used as a loss function.  The casting only occurs when the loss function is used as a metric.

The error is because of casting done in metrics.py on line 609.  See the snippet below from this file.

```
  def update_state(self, y_true, y_pred, sample_weight=None):
    """Accumulates metric statistics.

    `y_true` and `y_pred` should have the same shape.

    Args:
      y_true: Ground truth values. shape = `[batch_size, d0, .. dN]`.
      y_pred: The predicted values. shape = `[batch_size, d0, .. dN]`.
      sample_weight: Optional `sample_weight` acts as a
        coefficient for the metric. If a scalar is provided, then the metric is
        simply scaled by the given value. If `sample_weight` is a tensor of size
        `[batch_size]`, then the metric for each sample of the batch is rescaled
        by the corresponding element in the `sample_weight` vector. If the shape
        of `sample_weight` is `[batch_size, d0, .. dN-1]` (or can be broadcasted
        to this shape), then each metric element of `y_pred` is scaled by the
        corresponding value of `sample_weight`. (Note on `dN-1`: all metric
        functions reduce by 1 dimension, usually the last axis (-1)).

    Returns:
      Update op.
    """
    y_true = math_ops.cast(y_true, self._dtype)  # THIS IS WRONG!
    y_pred = math_ops.cast(y_pred, self._dtype)
    [y_true, y_pred], sample_weight = \
        metrics_utils.ragged_assert_compatible_and_get_flat_values(
            [y_true, y_pred], sample_weight)
    y_pred, y_true = losses_utils.squeeze_or_expand_dimensions(
        y_pred, y_true)

    ag_fn = autograph.tf_convert(self._fn, ag_ctx.control_status_ctx())
    matches = ag_fn(y_true, y_pred, **self._fn_kwargs)
    return super(MeanMetricWrapper, self).update_state(
        matches, sample_weight=sample_weight)
```

Setting a break point at the start of this function, we see the following vars in debug:
y_true
<tf.Tensor 'IteratorGetNext:3' shape=(None, 256, 256, 1) dtype=complex64>
self._dtype
'float32'

y_true should remain of type complex64 but instead is casted to float32 by ignoring the imaginary part and keeping the real part.

A correct solution will make an additional check for complex data in addition to the system floatx() type and cast accordingly (i.e. if floatx() is float32, complex128 becomes complex64; float64 goes to float32).

 I searched for this pattern,  tf.cast(y_true, self._dtype), and it occurs quite a bit throughout this file.  From what I can tell, all cases are incorrect and will cast complex data to float data and discard the imaginary part.  I also confirmed this bug exists in tf 2.7, reference https://github.com/keras-team/keras/blob/v2.7.0/keras/metrics.py#L1096-L1141

**Comment by sanatmpa1** [comment]
@isaacgerg,

Can you share a simple standalone code to reproduce the issue ? Thanks!

**Comment by isaacgerg** [comment]
@sanatmpa1 Unfortunately no.  Its self evident from the code i shared that complex numbers get casted to float and the imaginary part is discarded.

**Comment by mishc9** [comment]
@isaacgerg if you explicitly set metric's `dtype` to a complex type, it will use provided `dtype` for metic calculation as expected: for example, use `tf.keras.metrics.MeanSquaredError(dtype=tf.complex64)` instead of `'mse'` (automatic dtype casts could be helpful in this particular case, but made the framework more, er, complex. So probably it's not a bug). 
Does this solution work for you?

Example:

```python3
import tensorflow as tf
import numpy as np
from tensorflow.keras.layers import Dense


model = tf.keras.Sequential([Dense(32, 'relu'), Dense(1, 'sigmoid')])
model.compile(
    'sgd',
    loss=tf.keras.losses.MeanSquaredError(),
    metrics=[tf.keras.metrics.MeanSquaredError(dtype=tf.complex64)]
)

res = model.evaluate(
    np.random.normal(0, 1, [32, 64]),
    np.random.choice([0, 1], 32),
)

Out: 
1/1 [==============================] - 0s 212ms/step - loss: 0.3171 - mean_squared_error: 0.3171
/usr/local/lib/python3.7/dist-packages/keras/utils/generic_utils.py:960: ComplexWarning: Casting complex values to real discards the imaginary part
  info += ' %.4f' % avg

print(res)
Out:
[0.31706303358078003, (0.31706300377845764+0j)]   # or something like this... the loss value is real and the metic value is complex
# Inputs here are real, so the metric value on the real axis. It's just a demonstation
```
Complex dtypes are not allowed for some metrics - binary CE etc., and framework will complain about it. Also you cannot simply use standard losses with complex numbers, because they are not ordered, and the loss should 'define' some kind of order. That said, you probably would build custom loss and metric. That case you could set default value of [subclassed](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L147) metric to the complex dtype.

**Comment by isaacgerg** [comment]
@mishc9 I have a workaround but what you suggest will not generalize to my simple usecase. My metric essentially takes the normalmize(log(abs(complex_image))) of y_true and y_pred and then computes ssim on them.  Futhermore, as I show in the code, the metric framework ignores the complex datatype for general metric functions.  It works for yours because it is a special case.

**Comment by LukeWood** [comment]
It looks like the workaround provided should be sufficient, feel free to re-open if the workaround provided does not solve your use case.

**Comment by isaacgerg** [comment]
@LukeWood this is clearly a bug that you should fix.

**Comment by isaacgerg** [comment]
This is closed but is clearly a bug!

**Comment by LukeWood** [comment]
> @isaacgerg if you explicitly set metric's `dtype` to a complex type, it will use provided `dtype` for metic calculation as expected: for example, use `tf.keras.metrics.MeanSquaredError(dtype=tf.complex64)` instead of `'mse'` (automatic dtype casts could be helpful in this particular case, but made the framework more, er, complex. So probably it's not a bug). Does this solution work for you?
> 
> Example:
> 
> ```python
> import tensorflow as tf
> import numpy as np
> from tensorflow.keras.layers import Dense
> 
> 
> model = tf.keras.Sequential([Dense(32, 'relu'), Dense(1, 'sigmoid')])
> model.compile(
>     'sgd',
>     loss=tf.keras.losses.MeanSquaredError(),
>     metrics=[tf.keras.metrics.MeanSquaredError(dtype=tf.complex64)]
> )
> 
> res = model.evaluate(
>     np.random.normal(0, 1, [32, 64]),
>     np.random.choice([0, 1], 32),
> )
> 
> Out: 
> 1/1 [==============================] - 0s 212ms/step - loss: 0.3171 - mean_squared_error: 0.3171
> /usr/local/lib/python3.7/dist-packages/keras/utils/generic_utils.py:960: ComplexWarning: Casting complex values to real discards the imaginary part
>   info += ' %.4f' % avg
> 
> print(res)
> Out:
> [0.31706303358078003, (0.31706300377845764+0j)]   # or something like this... the loss value is real and the metic value is complex
> # Inputs here are real, so the metric value on the real axis. It's just a demonstation
> ```
> 
> Complex dtypes are not allowed for some metrics - binary CE etc., and framework will complain about it. Also you cannot simply use standard losses with complex numbers, because they are not ordered, and the loss should 'define' some kind of order. That said, you probably would build custom loss and metric. That case you could set default value of [subclassed](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L147) metric to the complex dtype.

@isaacgerg does explicitly setting `dtype` work?

**Comment by isaacgerg** [comment]
@LukeWood I am not sure.  I ended up rewriting the code to not have to depend on a user specified "type." Much cleaner and more mathematically correct in my opnion.

**Comment by LukeWood** [comment]
> @LukeWood I am not sure. I ended up rewriting the relevant keras code to not have to depend on a user specified "type" but use the given type. Much cleaner and more mathematically correct in my opnion.

Understandable.   I wonder if as a framework TensorFlow should raise a warning or error when a user casts a complex # to a real number.  Perhaps we should require an explicit call to `tf.real()`

**Comment by isaacgerg** [comment]
I can't comment on whether or not the bug belongs to tensorflow or keras.  I am simply a user reporting the bug (i think its my duty when using OSS). But, my team fixed it in our own in-house version of keras.  We never had to modify tensorflow.

The issue is that math_ops.cast() inside of keras does not properly handle complex types.  

For reference (since I have been using numpy for about 20 years):
>> np.array(np.complex(1,1)).astype('float32')
x-wingide-python-shell://124747328/7:1: ComplexWarning: Casting complex values to real discards the imaginary part
array(1., dtype=float32)

My expectation (which I am surely not alone in this) is that there would be a similar warning in tensorflow and if there isn't, keras would be aware that inputs could be complex and act mathematically correct.

Keras provides own casting library called math_ops as I am sure you are aware.
This library should handle complex data types correctly especially since they provide their own library on top of native tensorflow operations.

I know that many keras user likely don't have an idea why images would even be complex or assume if it is, that its complex throughout the network.  However, for many instances of remote sensing (my research area), this is not the case. I see you are Peter Gerstoff (sp?) student so your group should be quite familiar with these scenarios.

If I were at google, I would write unit tests to flush all of this starting with the behavior that complex types:
1. Issue a warning when they are truncated mathematically inconsistent as in the case of what numpy does for casting.  Note that np.real() and np.imag() give no such warnings -- it would make sense.
2. Correctly handle the management of complex data types since they are so abundant in real world data. This includes ensuring all the interfaces do not assume that data is real or if they do, they give an assertion.

Hope this helps.  Unfortunately, I cannot post any code due to restrictions.
